networkit/cpp/distance/DynDijkstra.cpp

```diff
@@ -10,26 +9,79 @@
 
 #include <networkit/auxiliary/Log.hpp>
 #include <networkit/auxiliary/NumericTools.hpp>
-#include <networkit/auxiliary/PrioQueue.hpp>
 #include <networkit/distance/Dijkstra.hpp>
 #include <networkit/distance/DynDijkstra.hpp>
 
-
 namespace NetworKit {
 
-DynDijkstra::DynDijkstra(const Graph& G, node source, bool storePredecessors) : DynSSSP(G, source, storePredecessors),
-color(G.upperNodeIdBound(), WHITE) {}
+DynDijkstra::DynDijkstra(const Graph &G, node source, bool storePredecessors)
+    : DynSSSP(G, source, storePredecessors), color(G.upperNodeIdBound(), WHITE),
+      heap(Aux::LessInVector<edgeweight>(distances)) {}
 
 void DynDijkstra::run() {
-    Dijkstra dij(*G, source, true);
-    dij.run();
-    distances = dij.getDistances();
-    npaths.reserve(G->upperNodeIdBound());
-    G->forNodes([&](node u) { npaths.push_back(dij.numberOfPaths(u)); });
+
+    color.clear();
+    count n = G->upperNodeIdBound();
+    color.resize(n, WHITE);
+
+    // init distances
+    distances.clear();
+    std::fill(distances.begin(), distances.end(), infDist);
+
+    if (distances.size() < G->upperNodeIdBound())
+        distances.resize(G->upperNodeIdBound(), infDist);
+
+    sumDist = 0.;
+    reachedNodes = 1;
+
     if (storePreds) {
+        previous.clear();
         previous.resize(G->upperNodeIdBound());
-        G->forNodes([&](node u) {previous[u] = dij.getPredecessors(u); });
     }
+
+    npaths.clear();
+    npaths.resize(G->upperNodeIdBound(), 0);
+    npaths[source] = 1;
+
+    // priority queue with distance-node pairs
+    distances[source] = 0.;
+    heap = tlx::d_ary_addressable_int_heap<node, 2, Aux::LessInVector<edgeweight>>(distances);
```

Comment:
This is not needed, I think it's better to do something like
```cpp
heap.clear();
heap.reserve(n);
```

Reply:
It is needed because the vector we are using the heap with can change (for example if we ran a dynamic update, but want then to rerun from scratch)

Reply:
We just need to make sure that the heap has no elements (so we call `clear()`) and that it has enough space to store all the nodes (so we call `reserve()`). We don't need to re-allocate a new heap (as we are doing now).